Earthfile

```diff
@@ -46,22 +47,31 @@ install-rust:
         --component clippy \
         --component rustfmt \
         --component llvm-tools-preview
-    RUN cargo install --locked --force --version 0.5.0 cargo-machete
-    RUN cargo install --locked --force --version 0.17.6 cargo-audit
     RUN cargo install --locked --force --version 0.36.11 cargo-make
     RUN cargo install --locked --force --version 0.5.22 cargo-llvm-cov
     RUN cargo install --locked --force --version 0.1.61 cargo-chef
+    RUN cargo install --locked --force --version 0.1.45 cargo-udeps
     RUN rustup --version
     RUN cargo --version
     RUN rustc --version
+    DO rust+INIT --keep_fingerprints=true
 
-formatting-check:
+rust-sources:
     FROM +install-rust
-    COPY Cargo.toml Cargo.toml
-    COPY rustfmt.toml rustfmt.toml
-    COPY crates/ crates/
-    COPY sql-to-dbsp-compiler/lib sql-to-dbsp-compiler/lib
-    RUN cargo +nightly fmt --all -- --check
+    COPY --keep-ts Cargo.toml Cargo.toml
+    COPY --keep-ts Cargo.lock Cargo.lock
+    COPY --keep-ts --dir crates crates
+    COPY --keep-ts sql-to-dbsp-compiler/lib sql-to-dbsp-compiler/lib
+    COPY --keep-ts README.md README.md
+
+formatting-check:
+    FROM +rust-sources
+    COPY --keep-ts rustfmt.toml rustfmt.toml
+    DO rust+CARGO --args="+nightly fmt --all -- --check"
+
+udeps:
```

Comment:
I think this will make the build worse because it has to compile everything twice, once nightly and once for non-nightly. Also at the time when I replaced it machete was supposedly better at finding unused dependencies than udeps?

Reply:
@gz We're already running nightly every time for `cargo fmt` though? Let me see if udeps works without nightly.

Reply:
Ah, actually machete catches unused dependencies in the sql libs, which might likely be accurate. I'll bring it back.

Reply:
I believe cargo fmt doesn't need to compile all dependencies but udeps does.. so urnning cargo +nightly udeps is a lot more expensive than cargo +nightly fmt

Reply:
IMO cargo machete can have some false positives but you can set tweak it by setting some meta-data in Cargo.toml

Reply:
It's working fine with machete now. I run it once for all projects without running it per crate.